File: Problem_3.py
"""
Time Complexity : O(n) 
Space Complexity : O(1)  
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No

The bruteforce could be to make all the subsets possible using backtracking , then multiply all of them and then add them. This
would be exponential again. To simplify this, this problem boild down to just a formulae.
Lets say we have [a,b,c]
All subsets- [a],[b],[c],[a,b][b,c],[a,c][a,b,c]
Our output should be a+b+c+ab+bc+ac+abc+1-1(we just add and subtract 1 for reaching the formula)
This boils down to (1+a) (1+b) (1+c) -1. So we just solve this and return. 
"""


def subsetProductSum(arr):
    result = 1
    for a in arr:
        result *= (a+1)
        print(result)
    return result - 1


arr = [1, 2, 3]
print(subsetProductSum(arr))

File: Problem3.java
/*
Sum of the products of all possible Subsets
approach:
An Efficient approach is to generalize the whole problem into some pattern. Suppose we have two numbers a and b. We can write all possible subset products as:-
   = a + b + ab
   = a(1+b) + b + 1 - 1
   = a(1+b) + (1+b) - 1
   = (a + 1) * (b + 1) - 1
   = (1+a) * (1 + b) - 1
Now take three numbers a, b, c:-
   = a + b + c + ab + bc + ca + abc
   = a + ac + b + bc + ab + abc + c + 1 - 1
   = a(1+c) + b(1+c) + ab(1+c) + c + 1 - 1
   = (a + b + ab + 1)(1+c) - 1
   = (1+a) * (1+b) * (1+c) - 1
time: O(n)
space: O(1)
 */
public class Problem3 {
    static int productOfSubsetSums(int arr[], int n) {
        int ans = 1;
        for (int i = 0; i < n; ++i)
            ans = ans * (arr[i] + 1);
        return ans - 1;
    }
}
